extensions/amp-3q-player/0.1/amp-3q-player.js

```diff
@@ -0,0 +1,190 @@
+ * @implements {../../../src/video-interface.VideoInterface}
+ */
+class Amp3QPlayer extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?Element} */
+    this.iframe_ = null;
+
+    /** @private {?Promise} */
+    this.playerReadyPromise_ = null;
+
+    /** @private {?Function} */
+    this.playerReadyResolver_ = null;
+  }
+
+  /**
+   * @param {boolean=} opt_onLayout
+   * @override
+   */
+  preconnectCallback(opt_onLayout) {
+    this.preconnect.url('https://playout.3qsdn.com', opt_onLayout);
+  }
+
+  /** @override */
+  buildCallback() {
+    this.playerReadyPromise_ = new Promise(resolve => {
+      this.playerReadyResolver_ = resolve;
+    });
+
+    const iframe = this.element.ownerDocument.createElement('iframe');
+    this.iframe_ = iframe;
+
+    this.forwardEvents([VideoEvents.PLAY, VideoEvents.PAUSE], iframe);
+    this.applyFillContent(iframe, true);
+    this.element.appendChild(iframe);
+
+    iframe.setAttribute('frameborder', '0');
+    iframe.setAttribute('allowfullscreen', 'true');
+
+    installVideoManagerForDoc(this.element);
+    videoManagerForDoc(this.element).register(this);
+  }
+
+  /** @override */
+  layoutCallback() {
+
+    const dataId = user().assert(
+        this.element.getAttribute('data-id'),
+        'The data-id attribute is required for <amp-3q-player> %s',
+        this.element);
+
+    const src = 'https://playout.3qsdn.com/' + dataId + '?autoplay=false&amp=true';
+    this.iframe_.src = src;
+
+    window.addEventListener('message',
+                            event => this.sdnBridge_(event));
+
+    return this.loadPromise(this.iframe_)
+      .then(() => {
+        this.element.dispatchCustomEvent(VideoEvents.LOAD);
+        this.playerReadyResolver_(this.iframe_);
+      });
+  }
+
+  /** @override */
+  unlayoutCallback() {
+    if (this.iframe_) {
+      removeElement(this.iframe_);
+      this.iframe_ = null;
+    }
+    return true;
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  viewportCallback(visible) {
+    this.element.dispatchCustomEvent(VideoEvents.VISIBILITY, {visible});
+  }
+
+  /** @override */
+  pauseCallback() {
+    if (this.iframe_) {
+      this.pause();
+    }
+  }
+
+  /** @private */
+  sdnBridge_(event) {
+    const data = isObject(event.data) ? event.data : tryParseJson(event.data);
+    if (data === undefined) {
+      return;
+    }
+
+    switch (data.data) {
+      case 'playing':
+        this.element.dispatchCustomEvent(VideoEvents.PLAY);
+        break;
+      case 'pause':
+        this.element.dispatchCustomEvent(VideoEvents.PAUSE);
+        break;
+
+    }
+  }
+
+  // VideoInterface Implementation. See ../src/video-interface.VideoInterface
+
+  /** @override */
+  play(unusedIsAutoplay) {
+    this.playerReadyPromise_.then(() => {
+      this.iframe_.contentWindow./*OK*/postMessage('play2', '*');
```

Comment:
none of these seems to work based on my testing. the postMessages for play, pause, mute and unmute seem to be ignored by the player. 

Please see the test page here: https://amphtml-ae5fc.firebaseapp.com/3q/examples/empty.amp.html